Civi/Core/CiviEventDispatcher.php

```diff
@@ -4,6 +4,8 @@
 
 use Symfony\Component\EventDispatcher\EventDispatcher;
 use Symfony\Component\EventDispatcher\Event;
+use Symfony\Component\HttpKernel\Kernel;
```

Comment:
@MikeyMJCO @totten this is annoying having to install the Kernel just to be able to test the Symfony version.

Reply:
Should be able to read the constant directly with:

```php
$symfony_version = \Symfony\Component\HttpKernel\Kernel::VERSION;
```

Reply:
It's about the same impact though and the `use` statement is probably more readable. It's not like we have to build a kernel or anything - we're just reading a constant.

Reply:
and I'm just catching up to it not being in `composer.json` before.

Reply:
yep

Reply:
I mean the kernel test isn't even reliable, it doesn't tell you the version of `EventDispatcher` which is what we need. You could, weirdly have a 4.4 kernel with a 3.x `EventDispatcher` - it'd be a weird edge case but it's possible since in 4.x Symfony components are independent with the Flex system.

Reply:
So if we could refactor this entirely to use `EventDispatcherInterface` we'd avoid the problem altogether since using `EventDispatcherInterface` is the same in 3.x and 4.x

Reply:
It's also worth noting that D8 goes EOL before D10 releases maybe we just say "screw it" and stop supporting D8 at that time and make the change to the 4.x style.

Reply:
The 8 -> 9 upgrade is pretty minor/tame. https://lab.civicrm.org/dev/core/-/issues/2316

Reply:
but I guess we still need some how to call dispatch() right and that is the problem because the function sig there changes so we would need to still know what version we are working with to know which order of the function params to pass through

Reply:
Could we use the LegacyDispatcher on Drupal 9.x? use Symfony\Component\EventDispatcher\LegacyEventDispatcherProxy;

Reply:
https://symfony.com/blog/new-in-symfony-4-3-simpler-event-dispatching

Reply:
So on Drupal 9 we'd do something like:
```php
public function __construct(EventDispatcherInterface $dispatcher)
    {
        $this->dispatcher = LegacyEventDispatcherProxy::decorate($dispatcher);
    }
 ```

Reply:
@MikeyMJCO so I see that as over working when all we care about is are we on symfony 4.3+ or not but there is no other version key other than in the http-kernel component

Reply:
I can see that, but the Kernel version doesn't in any way guarantee the version of the EventDispatcher.

Reply:
Trying to think through scenarios we're likely to see where it's plausible we could end up with a 4.x Kernel and 3.x dispatcher and they are incredibly unlikely. I think the `VERSION` check here is sufficient but that we should prioritise upgrading Symfony components in time for a D10 compatible release and articulate clearly a Drupal 8 support policy now.

Reply:
@totten any thoughts on whether you'd go with ending Drupal 8 support when D10 releases? We're looking at D8 going EoL in November 2021 and D10 being released in June 2022.

Reply:
Just saying we only stopped supporting D6 like 2 years ago I think

Reply:
Sure but that's a false equivalency. The D8 to D9 upgrade is in no way dramatic or problematic it's equivalent to a D6 or D7 minor version upgrade in most ways, I'm not suggesting that we immediately break compatibility in November, just that after November we're free to do so at any point - don't like it? Use ESR or ugprade to Drupal 9. Worth noting that the reason Drupal 8 goes EOL still in November is that's also EOL for Symfony 3.x something we're not preparing for either really.

Reply:
Regarding the version-check... Symfony 4.3+ made a change in the way contracts are managed for interfaces. 

* In <=4.2, `EventDispatcherInterface` is just a plain-old standalone interface.
* In >4.3, `EventDispatcherInterface` extends another interface (`Symfony\Contracts\EventDispatcher\EventDispatcherInterface` aka `ContractsEventDispatcherInterface`).

![image](https://user-images.githubusercontent.com/1336047/113696176-853bb980-9686-11eb-8ff7-dd840afaab53.png)

So maybe you could use `interface_exists` or `class_implements` to distinguish 4.2/4.3? (And the outcome should probably be cached? like in `$this` or `Civi::$statics`?)

Reply:
That's a good idea and more precise than checking HTTP Kernel versions.

Reply:
Does that work for Symfony v3 as well?

Reply:
so a check for `if ($eventDispatcherInterface->isSubClassOf('ContractsEventDispatcherInterface');` could work here?

Reply:
> Does that work for Symfony v3 as well?

It would return false since in Symfony 3 the class doesn't have the interface or parent (depending on which check we use!)

Reply:
You can also reformat as:

`is_subclass_of('EventDispatcherInterface','ContractsEventDispatcherInterface');`

Reply:
Which returns true if EventDispatcherInterface extends from ContractsEventDispatcherInterface or false otherwise.

Reply:
https://www.php.net/manual/en/function.is-subclass-of.php